SearchPhaseExecutionException:

The work in [#140515](https://github.com/elastic/elasticsearch/pull/140515) appears to throw an illegal state exception, which causes:

```
org.elasticsearch.action.search.SearchPhaseExecutionException: Failed to execute phase [query], all shards failed
```

And each shard has:

```
java.lang.IllegalStateException: Too many ids specified, allowed max result window is [10000]
```

Full stack trace:

```
      "type": "org.elasticsearch.action.search.SearchPhaseExecutionException"
Failed to execute phase [query], all shards failed; shardFailures {[FML043TOSu2fRY8f3YtkpQ][search-storypilot-nba-posts-blue][0]: org.elasticsearch.transport.RemoteTransportException: [es-es-search-76cf7b7d46-g7rp6][100.64.68.189:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.IllegalStateException: Too many ids specified, allowed max result window is [10000]
}{[mmNKh-I_TqegJvRU4AuCEQ][search-storypilot-nba-posts-blue][1]: java.lang.IllegalStateException: Too many ids specified, allowed max result window is [10000]
}{[FML043TOSu2fRY8f3YtkpQ][search-storypilot-nba-posts-blue][2]: org.elasticsearch.transport.RemoteTransportException: [es-es-search-76cf7b7d46-g7rp6][100.64.68.189:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.IllegalStateException: Too many ids specified, allowed max result window is [10000]
}{[FML043TOSu2fRY8f3YtkpQ][search-storypilot-nba-posts-blue][3]: org.elasticsearch.transport.RemoteTransportException: [es-es-search-76cf7b7d46-g7rp6][100.64.68.189:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.IllegalStateException: Too many ids specified, allowed max result window is [10000]
}{[mmNKh-I_TqegJvRU4AuCEQ][search-storypilot-nba-posts-blue][4]: java.lang.IllegalStateException: Too many ids specified, allowed max result window is [10000]
}{[mmNKh-I_TqegJvRU4AuCEQ][search-storypilot-nba-posts-blue][5]: java.lang.IllegalStateException: Too many ids specified, allowed max result window is [10000]
}
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:726)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.SearchQueryThenFetchAsyncAction.onPhaseFailure(SearchQueryThenFetchAsyncAction.java:84)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:340)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:772)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.finishOneShard(AbstractSearchAsyncAction.java:445)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:437)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.SearchQueryThenFetchAsyncAction.handleResponse(SearchQueryThenFetchAsyncAction.java:566)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.SearchQueryThenFetchAsyncAction.handleResponse(SearchQueryThenFetchAsyncAction.java:539)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.TransportService.handleResponse(TransportService.java:1750)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.TransportService.handleResponse(TransportService.java:1528)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.TransportService.handleResponse(TransportService.java:1528)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundHandler.doHandleResponse(InboundHandler.java:465)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundHandler.handleResponse(InboundHandler.java:419)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundHandler.executeResponseHandler(InboundHandler.java:166)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:134)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:102)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:831)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundPipeline.forwardFragment(InboundPipeline.java:116)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundPipeline.lambda(InboundPipeline.java:78)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundDecoder.finishMessage(InboundDecoder.java:144)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundDecoder.internalDecode(InboundDecoder.java:129)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundDecoder.decode(InboundDecoder.java:58)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:79)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:69)
	at org.elasticsearch.transport.netty4@9.5.0/org.elasticsearch.transport.netty4.Netty4MessageInboundHandler.channelRead(Netty4MessageInboundHandler.java:54)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at org.elasticsearch.transport.netty4@9.5.0/org.elasticsearch.transport.netty4.NettyByteBufSizer.channelRead(NettyByteBufSizer.java:33)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler@4.1.133.Final/io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1519)
	at io.netty.handler@4.1.133.Final/io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1377)
	at io.netty.handler@4.1.133.Final/io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1428)
	at io.netty.codec@4.1.133.Final/io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:545)
	at io.netty.codec@4.1.133.Final/io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:484)
	at io.netty.codec@4.1.133.Final/io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.transport@4.1.133.Final/io.netty.channel.DefaultChannelPipeline.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.transport@4.1.133.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.transport@4.1.133.Final/io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.transport@4.1.133.Final/io.netty.channel.nio.AbstractNioByteChannel.read(AbstractNioByteChannel.java:171)
	at io.netty.transport@4.1.133.Final/io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
	at io.netty.transport@4.1.133.Final/io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:697)
	at io.netty.transport@4.1.133.Final/io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:660)
	at io.netty.transport@4.1.133.Final/io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.common@4.1.133.Final/io.netty.util.concurrent.SingleThreadEventExecutor.run(SingleThreadEventExecutor.java:998)
	at io.netty.common@4.1.133.Final/io.netty.util.internal.ThreadExecutorMap.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1516)
Caused by: org.elasticsearch.ElasticsearchException: Too many ids specified, allowed max result window is [10000]
	at org.elasticsearch.server@9.5.0/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:811)
	at org.elasticsearch.server@9.5.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:338)
	... 50 more
Caused by: java.lang.IllegalStateException: Too many ids specified, allowed max result window is [10000]
```

--- a/docs/changelog/151234.yaml
+++ b/docs/changelog/151234.yaml
@@ -1,0 +1,6 @@
+area: Search
+issues:
+ - 150305
+pr: 151234
+summary: "Ids Query: Throw `IllegalArgumentException`"
+type: bug

--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/630_ids_query.yml
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/630_ids_query.yml
@@ -40,3 +40,44 @@ setup:
   - match: { error.root_cause.0.type: "query_shard_exception" }
   - match: { error.root_cause.0.reason: "failed to create query: Too many ids specified, allowed max result window is [2]" }
 
+---
+"ids query in aggregation filter returns error above max result window":
+  - requires:
+      cluster_features: ["ids_query_limit_max_ids"]
+      reason: "Ensure a limited number if ids can be supplied to ids query"
+
+  - do:
+      catch: bad_request
+      search:
+        body:
+          size: 0
+          aggs:
+            too_many_ids:
+              filter:
+                ids:
+                  values: [1, 2, 3]
+  - match: { status: 400 }
+  - match: { error.root_cause.0.type: "illegal_argument_exception" }
+  - match: { error.root_cause.0.reason: "Too many ids specified, allowed max result window is [2]" }
+
+---
+"ids query in highlight_query returns error above max result window":
+  - requires:
+      cluster_features: ["ids_query_limit_max_ids"]
+      reason: "Ensure a limited number if ids can be supplied to ids query"
+
+  - do:
+      catch: bad_request
+      search:
+        body:
+          query:
+            match_all: {}
+          highlight:
+            highlight_query:
+              ids:
+                values: [1, 2, 3]
+            fields:
+              key: {}
+  - match: { status: 400 }
+  - match: { error.root_cause.0.type: "illegal_argument_exception" }
+  - match: { error.root_cause.0.reason: "Too many ids specified, allowed max result window is [2]" }

--- a/server/src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java
+++ b/server/src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java
@@ -126,7 +126,7 @@ public class IdsQueryBuilder extends LeafQueryBuilder<IdsQueryBuilder> {
     @Override
     protected Query doToQuery(SearchExecutionContext context) throws IOException {
         if (ids.size() > context.indexSettings.getMaxResultWindow()) {
-            throw new IllegalStateException(
+            throw new IllegalArgumentException(
                 "Too many ids specified, allowed max result window is [" + context.indexSettings.getMaxResultWindow() + "]"
             );
         }

--- a/server/src/test/java/org/elasticsearch/index/query/IdsQueryBuilderTests.java
+++ b/server/src/test/java/org/elasticsearch/index/query/IdsQueryBuilderTests.java
@@ -101,7 +101,7 @@ public class IdsQueryBuilderTests extends AbstractQueryTestCase<IdsQueryBuilder>
             SearchExecutionContextHelper.SHARD_SEARCH_STATS
         );
 
-        IllegalStateException e = assertThrows(IllegalStateException.class, () -> builder.doToQuery(searchExecutionContext));
+        IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () -> builder.doToQuery(searchExecutionContext));
         assertThat(e.getMessage(), equalTo("Too many ids specified, allowed max result window is [2]"));
     }
 

